### ISSUE TITLE
Autodispose error prone checker

### ISSUE THREAD

**Issue by bangarharshit** [uber/AutoDispose#152]
AutoDispose error-prone checker can be really useful to catch lifecycle related errors in `Activity/Fragment`

[Ribs](https://github.com/uber/RIBs/blob/master/android/tooling/autodispose-error-prone-checker/src/main/java/com/uber/errorprone/checker/rx/RxJavaMissingAutodisposeErrorChecker.java) has a version of the checker. Why not make it generic to any object with a lifecycle?

Few scenarios I can think of

1. All the subclasses of standard android component such as `Activity/Fragment`
2. All the subclasses of `LifeCycleScopedProvider`
3. When the `LifeCycleScopedProvider` is composed instead of inheriting the standard component such as [Conductor](https://github.com/bluelinelabs/Conductor/tree/develop/conductor-modules/autodispose/src/main/java/com/bluelinelabs/conductor/autodispose)

**Comment by hzsweers** [comment]
Sure we could make an artifact. The only thing I'm hesitant about is that out internal implementation is a bit overbearing. Some synchronous streams should be allowed without binding. We could start with what we have and allow people to use it as an optional artifact. I would also prefer to have a lint artifact as well, so they'll show up in the IDE.

**Comment by bangarharshit** [comment]
Good point. `UAST` based lint checks can also be useful for Kotlin. 

The case 1 and 2 (standard Android components and subclasses of `LifeCycleScopedProvider`) should be trivial to fix. 

The case 3 where lifecycle is composed is pretty complex since the actual class (controller in this case) is not aware of anything about `LifeCycleScopedProvider` and may need a custom annotation or a marker interface `HasLifeCycle`.

We can start with checks for case 1 and 2. The check 3 and the case where it should be allowed without binding is something that can be picked up in a followup commit.

**Comment by bangarharshit** [comment]
Another useful thing is to have some kind of whitelisting mechanism. The base cases such as activity/fragment/LifecycleScopeProvider can be hardcoded in the test. 

We can use [error-prone flags](http://errorprone.info/docs/flags) and any library can configure it. For ribs -
```
XepOpt:AutoDisposeLeakCheck=com.uber.rib.core.Router,com.uber.rib.core.Presenter, com.uber.rib.core.Interactor, com.uber.rib.core.Worker 
```

**Comment by hzsweers** [comment]
Implementation is now in via #156, closing this is now pending on the wiki details

**Comment by bangarharshit** [comment]
Created a wiki in the PR - https://github.com/uber/AutoDispose/pull/162.  Do you think it makes sense to create a new task for lint rule?

**Comment by hzsweers** [comment]
This is done now!
